### ISSUE TITLE
Cannot write Variables with ENUM DataTypes to NetCDF-4

### ISSUE THREAD

**Issue by cwardgar** [Unidata/thredds#191]
From [KPZ-208087](https://andy.unidata.ucar.edu/esupport/staff/index.php?_m=tickets&_a=viewticket&ticketid=26109&departmentid=15&ticketstatusid=3)

Failure ultimately happens
`at ucar.nc2.jni.netcdf.Nc4Iosp.createVariable(Nc4Iosp.java:2395)`
where we try to call the `nc_def_var` NetCDF-4 C library method. The [docs](https://www.unidata.ucar.edu/software/netcdf/docs/netcdf-c/nc_005fdef_005fvar.html) say:
```
xtype
One of the set of predefined netCDF external data types. The type of this parameter,
nc_type, is defined in the netCDF header file. The valid netCDF external data types
are NC_BYTE, NC_CHAR, NC_SHORT, NC_INT, NC_FLOAT, and NC_DOUBLE. If the
file is a NetCDF-4/HDF5 file, the additional types NC_UBYTE, NC_USHORT, NC_UINT,
NC_INT64, NC_UINT64, and NC_STRING may be used, as well as a user defined type ID. 
```
but we're trying to create a variable with type `NC_ENUM`. Looking through Nc4Iosp as well as its tests, it doesn't look like we have any code at all to handle the enum case.

**Comment by cwardgar** [comment]
I added a failing test `ucar.nc2.jni.netcdf.TestNc4IospWriting.writeEnumType()` that demonstrates this issue: #192

**Comment by DennisHeimbigner** [comment]
I made some changes that make it work upto the point where it invokes
nc_var_puts. So, the problem is probably in the netcdf-c code. So I
am investigating that.

**Comment by DennisHeimbigner** [comment]
I did some additional testing where I duplicated
the testcase in c using the netcdf-c library directly.
It worked ok in that setting.
So, my current hypothesis is that you are using a
version of the netcdf-c library that is out-of-date.
At some point, nc_put_vars was changed to allow
it to work on user defined types (including enums).
Before that, it would fail. 
So, the thing I need to know is what version of the netcdf-c
library you are using. You can find this out as follows:
1. find the ncdump program associated with the library you are using.
2. Execute ncdump with no arguments. At the end of the usage output,
    it should say what library version is being used.

**Comment by cwardgar** [comment]
@DennisHeimbigner I'm on 4.3.2.

Does the `ucar.nc2.jni.netcdf.TestNc4IospWriting.writeEnumType()` fail on your machine?

**Comment by DennisHeimbigner** [comment]
Yes, but at a different place one I fixed some things.
=Dennis


Christian W wrote:

**Comment by DennisHeimbigner** [comment]
I have a fix on branch en.
The final problem was that I had a parameter
mixup and was not seeing it.

**Comment by cwardgar** [comment]
I improved the `writeEnumType()` test by adding a comparison of the in-memory NetcdfFile vs the one we write to disk: b63c717c39670c3a18a8a6cff751d1d24a279f85. It turned up another failure:
```
java.lang.NegativeArraySizeException
	at ucar.nc2.jni.netcdf.Nc4Iosp.readDataAll(Nc4Iosp.java:1727)
	at ucar.nc2.jni.netcdf.Nc4Iosp.readData(Nc4Iosp.java:1517)
	at ucar.nc2.NetcdfFile.readData(NetcdfFile.java:2026)
	at ucar.nc2.Variable.reallyRead(Variable.java:868)
	at ucar.nc2.Variable._read(Variable.java:839)
	at ucar.nc2.Variable.read(Variable.java:717)
	at ucar.nc2.util.CompareNetcdf2.compareVariableData(CompareNetcdf2.java:513)
	at ucar.nc2.util.CompareNetcdf2.compareVariables(CompareNetcdf2.java:309)
	at ucar.nc2.util.CompareNetcdf2.compareGroups(CompareNetcdf2.java:251)
	at ucar.nc2.util.CompareNetcdf2.compare(CompareNetcdf2.java:160)
	at ucar.nc2.jni.netcdf.TestNc4IospWriting.compare(TestNc4IospWriting.java:161)
```
I fixed this (I think) by using `en.getBaseType().getSize()` instead of `-1` for the size of the UserType (see my commit). While that stopped the exception from being thrown, the two NetcdfFile objects are still comparing unequal.

In-memory object:
```
netcdf null {
  types:
    short enum dessertType { 'pie' = 18, 'donut' = 268, 'cake' = 3284};

  dimensions:
    time = UNLIMITED;   // (3 currently)
  variables:
    enum dessertType dessert(time=3);

 data:
dessert =
  {18, 268, 3284}
}
```
On-disk object:
```
netcdf C:\Users\cwardgar\AppData\Local\Temp\writeEnumType5166302055763203471.nc {
  types:
    short enum dessertType { 'pie' = 18, 'donut' = 268, 'cake' = 3284};

  dimensions:
    time = UNLIMITED;   // (1 currently)
  variables:
    enum dessertType dessert(time=1);

 data:
dessert =
  {4608}
}
```
For some reason, the length of the `time` dimension is being changed from 3 to 1, and 4608 is a value in `dessert`, despite not even being in the enum map.

@DennisHeimbigner Any idea what's going on here?

**Comment by cwardgar** [comment]
Fixed by #228
